File: src/jvmMain/java/ess/papyrus/VarRef.java
package ess.papyrus;

import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;

/**
 * Variable that stores a ref. Note to self: a ref is a pointer to a papyrus
 * element, unlike a RefID which points to a form or changeform.
 */
final public class VarRef extends VarAbstractRef {

    public VarRef(ByteBuffer input, @NotNull PapyrusContext context) throws PapyrusFormatException {
        super(input, context);
    }

    public VarRef(TString type, EID id, @NotNull PapyrusContext context) {
        super(type, id, context);
    }

    @NotNull
    public VarRef derive(long id, @NotNull PapyrusContext context) {
        VarRef derivative = new VarRef(this.getRefType(), this.getRef().derive(id), context);
        return derivative;
    }

    @NotNull
    @Override
    public VarType getType() {
        return VarType.REF;
    }

}

File: src/jvmMain/kotlin/ess/papyrus/VarRef.kt
package ess.papyrus

import PlatformByteBuffer

/**
 * Variable that stores a ref. Note to self: a ref is a pointer to a papyrus
 * element, unlike a RefID which points to a form or changeform.
 */
class VarRef : VarAbstractRef {
    constructor(input: PlatformByteBuffer, context: PapyrusContext) : super(input!!, context) {}
    constructor(type: TString?, id: EID?, context: PapyrusContext) : super(type, id, context) {}

    fun derive(id: Long, context: PapyrusContext): VarRef {
        return VarRef(refType, ref.derive(id), context)
    }

    override val type: VarType
        get() = VarType.REF
}